### DISCUSSION TITLE
Proposal on the extension system for icon pack support

### DISCUSSION THREAD

**Discussion by SergioRibera** [zed-industries/zed#19488] [Feature Requests]
I wanted to propose an idea to improve the icon system as plugins. I was checking and there are already issues about it (https://github.com/zed-industries/zed/issues/8843), but I want to formalize it in this discussion and debate about the proposal in order to make a PR according to the needs.
This proposal was previously discussed in https://github.com/zed-industries/zed/issues/8843#issuecomment-2387676777

My proposal is that each language extension would provide a default icon and that there would be icon packs to replace these defaults. Zed would have a set of default icons that would act as a custom package, and the behavior would be as follows:

- If an icon package does not have an icon for a specific file type, the default icon of the language extension would be used.
- If neither the package nor the extension provides an icon, a generic icon would be displayed by default.
- To optimize performance, each time an extension is installed, a JSON file would be generated with the list of available icons (similar to how they already do with extensions).
- The idea is that if you have several icon packs installed you could select one from the selection menu.

This would be an example of how an `extension.toml` manifest for an extension that supports several languages might look like:

```toml
id = "cpp"
name = "C++"
version = "1.0.0"

[icons]
# custom icons by file type.
cpp = "icons/cpp.svg"
hpp = "icons/hpp.svg"
h = "icons/h.svg"
cxx = "icons/cxx.svg"
```

The JSON file generated to optimize access to the icons could look like this:

```json
{
    "h": "cpp/icons/h.svg",
    "cpp": "cpp/icons/cpp.svg",
    "hpp": "cpp/icons/hpp.svg",
    "cxx": "cpp/icons/cxx.svg"
}
```
    
In this way, the extension can manage its icons efficiently, without having to do too many unnecessary reads from the file system, improving performance.

If you find this proposal useful and would like to explore it further, I would be happy to collaborate with a prototype PR to test the idea.

> [!NOTE]
> This proposal aims to solve problems such as:
>
> What if I want to add a new language that does not come in the official Zed package? Will I have to wait for the Zed team to add the icon with its style? Where is the customisation? I mean I like flat icons, but someone else likes neon icons or more cartoon icons. I think this should be a bit more automatic and flexible, so my proposal goes that way.

**Comment by JosephTLyons** [comment]
Thanks for reaching out. We want to better support our community, and the best way for us to do that is through our official Discord. The team is active there and happy to help.

We’re closing the “Help and General Q&A” category on GitHub Discussions and directing all support questions to Discord going forward.

https://zed.dev/support
